src/universe/visualization/img_tools.py

```diff
@@ -14,3 +18,313 @@ def rotate(img: np.ndarray, angle: float) -> np.ndarray:
+            assert height > 0, 'Height must be greater than 0'
+            self.height = height
+
+    def scale(self, multiplier: float):
+        """
+        Scales the image by the given multiplier.
+
+        :param multiplier: multiplier to scale the image by
+        """
+        assert multiplier > 0, 'Multiplier must be greater than 0'
+        self.width = round(self.width * multiplier)
+        self.height = round(self.height * multiplier)
+
+
+def grayscale_image(image: cv.Mat) -> cv.Mat:
+    """
+    Converts the given image to grayscale and then to the original color space.
+
+    :param image: image to convert
+    :return: grayscale image
+    """
+    assert image is not None, 'Image does not exist'
+    return cv.cvtColor(cv.cvtColor(image, cv.COLOR_RGB2GRAY), cv.COLOR_GRAY2RGB)
+
+
+def darken_image(image: cv.Mat, darkness: int = 50) -> cv.Mat:
+    """
+    Darkens the given image.
+
+    :param image: image to darken
+    :param darkness: percentage of darkness (0-100)
+    :return: darkened image
+    """
+    assert image is not None, 'Image does not exist'
+    assert 0 <= darkness <= 100, 'Darkness percentage must be between 0 and 100'
+    return cv.addWeighted(np.zeros_like(image), darkness / 100, image, 1 - darkness / 100, 0)
+
+
+def draw_lines(image: cv.Mat, lines: List[Line2d], color: Tuple[int, int, int],
+               endpoints: bool = False, line_width: float = 0.004, opacity: int = 75) -> cv.Mat:
+    """
+    Draws the given lines on the given image with the given color.
+    Make sure that the lines are scaled to the image size.
+
+    :param image: image to draw the lines on
+    :param lines: lines to draw on the image (scaled to the image size)
+    :param color: color of the lines (in the same color space as the image)
+    :param endpoints: whether to draw circles at the endpoints of the lines
+    :param line_width: this value multiplied by the width of the image is the line width in pixels
+    :param opacity: opacity of the lines (0-100)
+    :return: image with the lines drawn on it
+    """
+    assert image is not None, 'Image does not exist'
+    assert lines and len(lines) != 0, 'No lines to draw'
+    assert 0 <= opacity <= 100, 'Opacity must be between 0 and 100'
+
+    # calculate the line width in pixels based on the image width
+    line_width = round(line_width * image.shape[1])
+
+    # create an overlay for the lines
+    overlay = np.zeros_like(image)
+    for line in lines:
+        # round the points to integers
+        p1 = (round(line.a.x), round(line.a.y))
+        p2 = (round(line.b.x), round(line.b.y))
+
+        # draw line on the overlay
+        cv.line(overlay, p1, p2, color, line_width)
+        # draw endpoints on the overlay
+        if endpoints:
+            cv.circle(overlay, p1, line_width, color, -1)
+            cv.circle(overlay, p2, line_width, color, -1)
+
+    # draw the overlay on the image
+    return cv.addWeighted(overlay, opacity / 100, image, 1., 0.)
```

Comment:
Same as earlier: opacity between 0 and 1. Also, shouldn't it be 1. - opacity ?
```suggestion
    return cv.addWeighted(overlay, opacity, image, 1. - opacity, 0.)
```

Reply:
No, because then the image would get darker when drawing the layer

Reply:
make sense.